### PR TITLE
Add: tensorflow example for image classification task guide

### PR DESCRIPTION
**Pull request by MKhalusova**
This PR addresses https://github.com/huggingface/transformers/issues/21037
It adds a Tensorflow example to the existing task guide on image classification. 

State of the PR: 
The example illustrates preprocessing in TF, training, and pushing to Hub. The code samples have been tested and they work/reproduce.

### CODE REVIEW

File: docs/source/en/tasks/image_classification.mdx [hunk 1]
```diff
@@ -228,6 +276,144 @@ Once training is completed, share your model to the Hub with the [`~transformers
...
+...     columns=["pixel_values"], label_cols=["label"], shuffle=True, batch_size=batch_size, collate_fn=data_collator
+... )
+
+>>> # converting our test dataset to tf.data.Dataset
+>>> tf_eval_dataset = food["test"].to_tf_dataset(
+...     columns=["pixel_values"], label_cols=["label"], shuffle=True, batch_size=batch_size, collate_fn=data_collator
+... )
+```
+
+Configure the model for training with `compile()`:
```
Comment (amyeroberts) [inline docs/source/en/tasks/image_classification.mdx:340]: nit: possible suggestion, feel free to ignore. To keep spacing consistent (one new line before and after code block)

```suggestion
Configure the model for training with `compile()`:

```

File: docs/source/en/tasks/image_classification.mdx [hunk 2]
```diff
@@ -228,6 +276,144 @@ Once training is completed, share your model to the Hub with the [`~transformers
...
+>>> api.upload_folder(
+...     token=hub_token,
+...     folder_path=output_dir,
+...     path_in_repo="",
+...     repo_id=f"MariaK/{output_dir}",
+...     ignore_patterns="**/logs/*.txt",
+... )
+```
+
+Congratulations! Your model has been shared on 🤗Hub, and you can use it for inference!
```
Comment (amyeroberts) [inline docs/source/en/tasks/image_classification.mdx:412]: nit: same as above re spacing 

```suggestion
Congratulations! Your model has been shared on 🤗 Hub, and you can use it for inference!
```

**Review by amyeroberts — COMMENTED** [review]
Very nice! Thanks for adding this example 🤗 

Only main comment I have is a question about the changing outputs from the model in the example. Other than that, it's all just ultra nits. 

I'll do a final re-review once it's not longer WIP but looks ready to merge as-is to me.

File: docs/source/en/tasks/image_classification.mdx [hunk 3]
```diff
@@ -228,6 +276,144 @@ Once training is completed, share your model to the Hub with the [`~transformers
 </pt>
 </frameworkcontent>
 
+<frameworkcontent>
+<tf>
+
+<Tip>
+If you are unfamiliar with fine-tuning a model with Keras, check out the <a href="https://huggingface.co/docs/transformers/main/en/training#train-a-tensorflow-model-with-keras">basic tutorial</a> first!
```
[the remark this answers is not in the document]
Comment (MKhalusova) [reply docs/source/en/tasks/image_classification.mdx:283]: AFAIK, there's an issue with rendering markdown links inside a <tip>

File: docs/source/en/tasks/image_classification.mdx [hunk 4]
```diff
@@ -130,17 +136,57 @@ To apply the preprocessing function over the entire dataset, use 🤗 Datasets [
...
+<tf>
+Create a preprocessing function to resize the images and return the `pixel_values`, which are required inputs to the model.
+
+```py
+>>> def process(examples):
+...     examples.update(
+...         image_processor(
+...             examples["image"],
+...         )
+...     )
```
[the remark this answers is not in the document]
Comment (MKhalusova) [reply docs/source/en/tasks/image_classification.mdx:161]: I had it in one line initially but `make style` converted it this way. If I revert it back it'll likely fail on the CI side.

File: docs/source/en/tasks/image_classification.mdx [hunk 5]
```diff
@@ -228,6 +276,144 @@ Once training is completed, share your model to the Hub with the [`~transformers
...
+... )
+```
+
+Then, load ViT with `TFAutoModelForImageClassification` along with the label mappings:
+
+```py
+>>> from transformers import TFAutoModelForImageClassification
+
+>>> model = TFAutoModelForImageClassification.from_pretrained(
+...     checkpoint,
```
[the remark this answers is not in the document]
Comment (MKhalusova) [reply docs/source/en/tasks/image_classification.mdx:320]: good point! I'll modify the example

File: docs/source/en/tasks/image_classification.mdx (hunk 4 above)
[the remark this answers is not in the document]
Comment (amyeroberts) [reply docs/source/en/tasks/image_classification.mdx:160]: I think `make style` made it across several lines because `black` will [split across many lines if there's a comma](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#the-magic-trailing-comma) and no element after it e.g. 

```python
foo = bar(a, b, c,)
```

will be set to 

```python
foo = bar(
    a,
    b,
    c,
)
```

In the suggestion I posted above it should allow you to leave it on one line.

File: docs/source/en/tasks/image_classification.mdx [hunk 6]
```diff
@@ -256,12 +442,13 @@ The simplest way to try out your finetuned model for inference is to use it in a
 
 >>> classifier = pipeline("image-classification", model="my_awesome_food_model")
 >>> classifier(image)
-[{'score': 0.35574808716773987, 'label': 'beignets'},
- {'score': 0.018057454377412796, 'label': 'chicken_wings'},
- {'score': 0.017733804881572723, 'label': 'prime_rib'},
- {'score': 0.016335085034370422, 'label': 'bruschetta'},
- {'score': 0.0160061065107584, 'label': 'ramen'}]
-```
+[{'score': 0.9209485054016113, 'label': 'beignets'},
```
[the remark this answers is not in the document]
Comment (MKhalusova) [reply docs/source/en/tasks/image_classification.mdx:438]: I replaced the initial values with the values that I got when I use my trained model in this pipeline. However, I should've then replaced the `model` parameter value with the id of my model. The original values do seem rather low actually. I feel like this model should be more confident. I'll retrain the pytorch version exactly as in the tutorial to see if they are really that low.

File: docs/source/en/tasks/image_classification.mdx [hunk 7]
```diff
@@ -155,21 +194,23 @@ Then create a function that passes your predictions and labels to [`~evaluate.Ev
...
+Your `compute_metrics` function is ready to go now, and you'll return to it when you set up your training.
 
 ## Train
 
 <frameworkcontent>
 <pt>
 <Tip>
 
-If you aren't familiar with finetuning a model with the [`Trainer`], take a look at the basic tutorial [here](../training#train-with-pytorch-trainer)!
+If you aren't familiar with fine-tuning a model with the [`Trainer`], take a look at the basic tutorial [here](../training#train-with-pytorch-trainer)!
```
Comment (sgugger) [inline docs/source/en/tasks/image_classification.mdx:210]: I have said it before, there is no standard for finetune vs fine-tune in the library and we won't enforce one. While you're free to pick the one you like when writing new content, please leave the existing ones as is so that the diff is easier to read for the reviewer :-)

File: docs/source/en/tasks/image_classification.mdx (hunk 3 above)
[the remark this answers is not in the document]
Comment (sgugger) [reply docs/source/en/tasks/image_classification.mdx:276]: Even when having new lines after <Tip> and before </Tip>? I didn't see another example in the lib but the links in `DataCollatorForLanguageModeling` to other objects in the doc are resolved to Markdown links and seem to render normally [here](https://huggingface.co/docs/transformers/main_classes/data_collator#transformers.DataCollatorForLanguageModeling).

File: docs/source/en/tasks/image_classification.mdx [hunk 8]
```diff
@@ -228,6 +269,144 @@ Once training is completed, share your model to the Hub with the [`~transformers
...
+
+>>> optimizer, lr_schedule = create_optimizer(
+...     init_lr=learning_rate,
+...     num_train_steps=num_train_steps,
+...     weight_decay_rate=weight_decay_rate,
+...     num_warmup_steps=0,
+... )
+```
+
+Then, load ViT with `TFAutoModelForImageClassification` along with the label mappings:
```
Comment (sgugger) [inline docs/source/en/tasks/image_classification.mdx:307]:
```suggestion
Then, load ViT with [`TFAutoModelForImageClassification`] along with the label mappings:
```
To link to the doc of this class.

File: docs/source/en/tasks/image_classification.mdx [hunk 9]
```diff
@@ -228,6 +269,144 @@ Once training is completed, share your model to the Hub with the [`~transformers
...
+>>> from transformers import TFAutoModelForImageClassification
+
+>>> model = TFAutoModelForImageClassification.from_pretrained(
+...     checkpoint,
+...     id2label=id2label,
+...     label2id=label2id,
+... )
+```
+
+Convert your datasets to the `tf.data.Dataset` format using the [`~datasets.Dataset.to_tf_dataset()`] and your `data_collator`:
```
Comment (sgugger) [inline docs/source/en/tasks/image_classification.mdx:319]:
```suggestion
Convert your datasets to the `tf.data.Dataset` format using the [`~datasets.Dataset.to_tf_dataset`] and your `data_collator`:
```
The parenthesis will be added automatically (but putting them will prevent the doc-builder from finding that method as you can see in the preview).

File: docs/source/en/tasks/image_classification.mdx [hunk 10]
```diff
@@ -228,6 +269,144 @@ Once training is completed, share your model to the Hub with the [`~transformers
...
+>>> model.save_pretrained(output_dir)
+
+>>> api.upload_folder(
+...     token=hub_token,
+...     folder_path=output_dir,
+...     path_in_repo="",
+...     repo_id=f"MariaK/{output_dir}",
+...     ignore_patterns="**/logs/*.txt",
+... )
+```
```
Comment (sgugger) [inline docs/source/en/tasks/image_classification.mdx:403]: This is way too complicated compared to using the `PushToHubCallback`. Why not use it? The only weird thing is that the `image_processor` should be passed to the `tokenizer` argument (which might need to be renamed to `processor` at some point).

**Review by sgugger — COMMENTED** [review]
Thanks for working on the TensorFlow side of this guide! I left a couple of comments, the main one is on the push-to-hub part.

File: docs/source/en/tasks/image_classification.mdx (hunk 10 above)
Comment (MKhalusova) [reply docs/source/en/tasks/image_classification.mdx:403]: This is the reason why I am not using `PushToHubCallback` at the moment, it hangs the cell execution. https://github.com/huggingface/transformers/issues/21116

Comment (sgugger) [reply docs/source/en/tasks/image_classification.mdx:407]: Let's remember to move back to it when the bug is fixed then :-)

File: docs/source/en/tasks/image_classification.mdx (hunk 3 above)
[the remark this answers is not in the document]
Comment (MKhalusova) [reply docs/source/en/tasks/image_classification.mdx]: Fixed in 39a13a90375e52a9e373bf71d01309bc80531578

File: docs/source/en/tasks/image_classification.mdx [hunk 11]
```diff
@@ -228,6 +276,144 @@ Once training is completed, share your model to the Hub with the [`~transformers
...
 
+<frameworkcontent>
+<tf>
+
+<Tip>
+If you are unfamiliar with fine-tuning a model with Keras, check out the <a href="https://huggingface.co/docs/transformers/main/en/training#train-a-tensorflow-model-with-keras">basic tutorial</a> first!
+</Tip>
+
+To fine-tune a model in TensorFlow, follow these steps:
+1. define the training hyperparameters, and set up an optimizer and a learning rate schedule
```
[the remark this answers is not in the document]
Comment (MKhalusova) [reply docs/source/en/tasks/image_classification.mdx]: Fixed in 39a13a90375e52a9e373bf71d01309bc80531578

File: docs/source/en/tasks/image_classification.mdx (hunk 5 above)
[the remark this answers is not in the document]
Comment (MKhalusova) [reply docs/source/en/tasks/image_classification.mdx:316]: 39a13a90375e52a9e373bf71d01309bc80531578

File: docs/source/en/tasks/image_classification.mdx (hunk 1 above)
Comment (MKhalusova) [reply docs/source/en/tasks/image_classification.mdx:336]: Fixed in 39a13a90375e52a9e373bf71d01309bc80531578

File: docs/source/en/tasks/image_classification.mdx (hunk 6 above)
[the remark this answers is not in the document]
Comment (MKhalusova) [reply docs/source/en/tasks/image_classification.mdx]: I re-trained the pytorch model exactly as is in the tutorial, and you're right, the numbers are similar to what they were originally, so I reverted back to them in 39a13a90375e52a9e373bf71d01309bc80531578

File: docs/source/en/tasks/image_classification.mdx (hunk 8 above)
Comment (MKhalusova) [reply docs/source/en/tasks/image_classification.mdx]: Fixed in 39a13a90375e52a9e373bf71d01309bc80531578

File: docs/source/en/tasks/image_classification.mdx (hunk 9 above)
Comment (MKhalusova) [reply docs/source/en/tasks/image_classification.mdx]: Fixed in 39a13a90375e52a9e373bf71d01309bc80531578

File: docs/source/en/tasks/image_classification.mdx (hunk 10 above)
Comment (MKhalusova) [reply docs/source/en/tasks/image_classification.mdx]: So according to the feedback on the issue, `PushToHubCallback` partly works. If save_strategy is set to "no", it will push the model on the train completion, so I switched to the callback. 
Once the issue is resolved I can remove the save_strategy parameter. Please review. 63f1d9fdfc546c40f1e3cfbcb8e29f9a5787ae1f

File: docs/source/en/tasks/image_classification.mdx [hunk 12]
```diff
@@ -228,6 +270,116 @@ Once training is completed, share your model to the Hub with the [`~transformers
...
+>>> loss = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
+>>> model.compile(optimizer=optimizer, loss=loss)
+```
+
+To compute the accuracy from the predictions and push your model to the 🤗 Hub, use [Keras callbacks](./main_classes/keras_callbacks).
+Pass your `compute_metrics` function to [KerasMetricCallback](./main_classes/keras_callbacks#transformers.KerasMetricCallback),
+and use the [PushToHubCallback](./main_classes/keras_callbacks#transformers.PushToHubCallback) to upload the model:
+
+```py
+>>> from transformers.keras_callbacks import KerasMetricCallback
```
Comment (MKhalusova) [inline docs/source/en/tasks/image_classification.mdx:350]:
```suggestion
>>> from transformers.keras_callbacks import KerasMetricCallback, PushToHubCallback
```
```suggestion
>>> from transformers.keras_callbacks import KerasMetricCallback
```

**Review by sgugger — APPROVED** [review]
Thanks again for your work on this!

**Comment by MKhalusova** [comment]
Thank you all for the reviews! 
I have added the final part - data augmentation (kudos to @sayakpaul for helping me troubleshoot the issues I was having). The example is now complete. Let me know if it looks good enough to be merged :) 
cc @amyeroberts @sgugger

**Review by sgugger — APPROVED** [review]
Yes, still good for me!

**Review by amyeroberts — APPROVED** [review]
All looks good to me! Thanks for adding this guide
